File: rust/src/bin/day5.rs
use std::{
    collections::{HashMap, VecDeque},
    fs,
};

use itertools::Itertools;
use regex::Regex;

lazy_static::lazy_static! {
    static ref FILE:String = fs::read_to_string("../input/day5.txt").unwrap();
}

fn main() {
    print!("{}{:?}", "Day 5 ", day5(&FILE));
}

fn day5(input: &str) -> Vec<String> {
    let sections = input.split("\n\n").collect_vec();
    let mut layers = sections[0].split("\n").collect();
    let mut stacks = process_stack(&mut layers);
    let moves = sections[1].split("\n").collect_vec();
    let part1 = execute_moves(&mut stacks.clone(), &moves, false);
    let part2 = execute_moves(&mut stacks, &moves, true);
    vec![part1, part2]
}

fn process_stack(layers: &mut VecDeque<&str>) -> HashMap<u32, VecDeque<char>> {
    let mut stacks = HashMap::new();
    let mut labels = HashMap::new();
    for (idx, label) in layers.pop_back().unwrap().chars().enumerate() {
        if label != ' ' {
            labels.insert(label.to_digit(10).unwrap(), idx);
        }
    }
    while let Some(layer) = layers.pop_back() {
        for (label, idx) in labels.iter() {
            let stack = stacks.entry(*label).or_insert_with(VecDeque::new);
            stack.extend(layer.chars().nth(*idx).filter(|&ch| ch != ' '));
        }
    }
    stacks
}

fn execute_moves(stacks: &mut HashMap<u32, VecDeque<char>>, moves: &Vec<&str>, all_at_once: bool) -> String {
    let mut top = String::from("");
    let re = Regex::new(r"move (\d+) from (\d+) to (\d+)").unwrap();
    for line in moves {
        let numbers: Vec<_> = re.captures(line).unwrap().iter().skip(1).map(|num| num.unwrap().as_str().parse::<u32>().unwrap()).collect();
        let mut temp_stack = VecDeque::new();
        for _ in 0..numbers[0] {
            if all_at_once {
                if let Some(item) = stacks.get_mut(&numbers[1]).and_then(|src| src.pop_back()) {
                    temp_stack.push_front(item);
                }
            } else {
                stacks.get_mut(&numbers[1]).and_then(|src| src.pop_back()).and_then(|item| stacks.get_mut(&numbers[2]).map(|dest| dest.push_back(item)));
            }
        }
        stacks.get_mut(&numbers[2]).unwrap().extend(temp_stack);
    }
    for key in 1..stacks.len() + 1 {
        top.push(stacks.get_mut(&(key as u32)).unwrap().pop_back().unwrap());
    }
    top.to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn day5_part1_test() {
        assert_eq!(day5(&FILE), vec!("BWNCQRMDB", "NHWZCBNBF"))
    }
}

File: py/Day5.py
from collections import deque

def read_file(file):
    layers, moves, process_layers = [], [], True
    with open(file) as file:
        for line in file.readlines():
            if line == "\n":
                process_layers = False
            else:
                (layers if process_layers else moves).append(line.strip("\n"))
    return layers, moves

def process_stacks():
    labels = layers.pop()
    index = {int(label): idx for idx, label in enumerate(labels) if label != " "}
    stacks = [deque() for _ in range(int(labels[len(labels) - 1]) + 1)]
    for stack in layers[::-1]:
        for label, idx in index.items():
            if idx < len(stack) and stack[idx] != " ":
                stacks[label].append(stack[idx])
    return stacks

def execute_moves(stacks, all_at_once):
    for move in moves:
        nums = [int(str) for str in move.split() if str.isdigit()]
        times, src, dest, temp_stacks = nums[0], nums[1], nums[2], []
        for _ in range(times):
            temp_stacks.append(stacks[src].pop()) if all_at_once else stacks[dest].append(stacks[src].pop())
        stacks[dest].extend(reversed(temp_stacks))  # part 2
    return stacks

def get_top_layers(stacks):
    return "".join(stack.pop() if stack else "" for stack in stacks)

layers, moves = read_file('../input/day5.txt')
list_stacks = process_stacks()
print("Part I: ", get_top_layers(execute_moves([deque(stack) for stack in list_stacks], False)))
print("Part II:", get_top_layers(execute_moves(list_stacks, True)))
